File: scratch_wall/scratch_wall.js

var box;
var remote ="";
var pending = false;
var diff = new diff_match_patch();

window.onload = function() {
    box = document.getElementById('scratch');
    box.onkeyup = function(event){ pending = true; };
}

function push() {
    if(pending){
	var xmlHttpReq = false;
	var self = this;
	// Mozilla/Safari
	if (window.XMLHttpRequest) {
	    self.xmlHttpReq = new XMLHttpRequest();
	}
	// IE
	else if (window.ActiveXObject) {
	    self.xmlHttpReq = new ActiveXObject("Microsoft.XMLHTTP");
	}
	self.xmlHttpReq.open('POST', "scratch_wall.py/push", true);
	self.xmlHttpReq.setRequestHeader('Content-Type',
					 'application/x-www-form-urlencoded');
	self.xmlHttpReq.onreadystatechange = function() {
	    if (self.xmlHttpReq.readyState == 4) {
	    }
	}
	var to_send = encodeURIComponent(diff.patch_toText(diff.patch_make(remote,box.value)));
	self.xmlHttpReq.send("delta="+to_send);
	remote=box.value;
	pending = false;
    }
    setTimeout("push()",1000);
}

function pull(){
    var xmlHttpReq = false;
    var self = this;
    // Mozilla/Safari
    if (window.XMLHttpRequest) {
	self.xmlHttpReq = new XMLHttpRequest();
    }
    // IE
    else if (window.ActiveXObject) {
	self.xmlHttpReq = new ActiveXObject("Microsoft.XMLHTTP");
    }
    self.xmlHttpReq.open('GET', "scratch_wall.py/pull", true);
    self.xmlHttpReq.onreadystatechange = function() {
	if (self.xmlHttpReq.readyState == 4) {
	    var patch = self.xmlHttpReq.responseText;    
	    remote=diff.patch_apply(diff.patch_fromText(patch),remote)[0];
	    box.value=diff.patch_apply(diff.patch_make(box.value,remote),box.value)[0];
	    box.value = remote;
	}
    }
    self.xmlHttpReq.send();
    setTimeout("pull()",1000);
    
}

push();
pull();

File: scratch_wall/scratch_wall.py
#!/usr/bin/python
from diff_match_patch import diff_match_patch, patch_obj
from mod_python import apache
from mod_python import Session
import sys
diff = diff_match_patch();

def index(req, wall=""):	
	if(wall == ''):
		req.headers_out['location'] = 'http://172.21.21.39/scratch_wall.py?wall=random'
		req.status = apache.HTTP_MOVED_TEMPORARILY
		return	
	session = Session.Session(req)
	#try:
	session['wall'] = wall
	session['updates']=1
	session['text'] =''

	open('/home/josh/scratch_wall/walls/%s' % session['wall'],'a').close()
	req.content_type = "text/html"
	with open('/home/josh/scratch_wall/scratch_wall.html', 'r') as main:
		req.write(main.read())
	session.save()
	

def pull(req):
	session = Session.Session(req)
	patch = ''
	with open('/home/josh/scratch_wall/walls/%s' % session['wall'],'r') as wall:
		session['updates'] += 1
		text = wall.read()
		patch = diff.patch_toText(diff.patch_make(session['text'],text))	
		session['text'] = text		
	session.save()
	req.write( patch )
	

def push(req):
	session = Session.Session(req)
	session['updates'] = 1
	session['text'] = diff.patch_apply(diff.patch_fromText(req.form['delta']),session['text'])[0]
	with open('/home/josh/scratch_wall/walls/%s' % session['wall'],'w') as wall:
		wall.write( session['text']);
	session.save()





